File: Algorithm_Swift/Algorithm_Swift/Programmers/Programmers_StringSort.swift
//
//  Programmers_StringSort.swift
//  Algorithm_Swift
//
//  Created by asong on 2022/03/02.
//

import Foundation

func solution(_ strings:[String], _ n:Int) -> [String] {
    var answer: [String] = []
    var indexList: [String] = []
    var dict: [String: Array<String>] = [:]
    for i in 0..<strings.count {
        let str = strings[i]
        let s = String(str[str.index(str.startIndex, offsetBy: n)])
        if !indexList.contains(s) {
            indexList.append(s)
            dict[s] = Array<String>()
        }
        var arr: [String] = dict[s]!
        arr.append(strings[i])
        dict[s] = arr
    }
    
    indexList.sort()
    for c in indexList {
        var newList: [String] = dict[c]!
        newList.sort()
        for str in newList {
            answer.append(str)
        }
    }
    return answer
}

File: Econovation_Algorithm_Mentoring/2022코테준비/Programmers/Programmers_StringSort.java
import java.util.*;

public class Programmers_StringSort {
        public ArrayList<String> solution(String[] strings, int n) {

            ArrayList<Character> list = new ArrayList<>();
            HashMap<Character, ArrayList<String>> hm = new HashMap<>();
            ArrayList<String> answer = new ArrayList<>();

            for(int i=0;i<strings.length;i++){
                char c = strings[i].charAt(n);
                if(!list.contains(c)){
                    list.add(c);
                    hm.put(c, new ArrayList<String>());
                }
                ArrayList<String> llist = hm.get(c);
                llist.add(strings[i]);
                hm.put(c, llist);
            }
            Collections.sort(list);
            for(int i=0;i<list.size();i++){
                char c = list.get(i);
                ArrayList<String> newList = hm.get(c);
                Collections.sort(newList);
                for(int j=0;j<newList.size();j++){
                    answer.add(newList.get(j));
                }
            }

            return answer;
        }
}
